core/common/src/main/java/alluxio/master/PortRegistry.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.net.ServerSocket;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.channels.OverlappingFileLockException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Registry for reserving ports during tests.
+ *
+ * The registry reserves ports by taking file locks on files in the port coordination directory.
+ * This doesn't prevent external processes from stealing our ports, but it will prevent us from
+ * conflicting with ourselves. We can then run tests in a dockerized environment to completely
+ * prevent conflicts.
+ *
+ * The default coordination directory is determined by the "user.dir" jvm property. The coordination
+ * directory can be overridden by setting the ALLUXIO_PORT_COORDINATION_DIR environment variable.
+ */
+public final class PortRegistry {
+  private static final String PORT_COORDINATION_DIR_PROPERTY = "ALLUXIO_PORT_COORDINATION_DIR";
+
+  public static final PortRegistry INSTANCE = new PortRegistry();
+
+  // Map from port number to the reservation for that port.
+  private final Map<Integer, Reservation> mReserved = new ConcurrentHashMap<>();
+  private final File mCoordinationDir;
+
+  @VisibleForTesting
+  PortRegistry() {
+    String dir = System.getenv(PORT_COORDINATION_DIR_PROPERTY);
+    if (dir == null) {
+      dir = System.getProperty("user.dir");
+    }
+    mCoordinationDir = new File(dir, ".port_coordination");
+    mCoordinationDir.mkdirs();
+  }
+
+  /**
+   * Reserves a free port so that other tests will not take it.
+   *
+   * @return the free port
+   */
+  public int reservePort() {
+    for (int i = 0; i < 1000; i++) {
+      int port = getFreePort();
+      if (lockPort(port)) {
+        return port;
+      }
+    }
+    throw new RuntimeException("Failed to acquire port");
+  }
+
+  /**
+   * Attempts to lock the given port.
+   *
+   * @param port the port to lock
+   * @return whether the locking succeeded
+   */
+  @VisibleForTesting
+  boolean lockPort(int port) {
+    File portFile = portFile(port);
+    try {
+      FileChannel channel = new RandomAccessFile(portFile, "rw").getChannel();
+      FileLock lock = channel.tryLock();
+      if (lock == null) {
+        channel.close();
+        return false;
+      }
+      mReserved.put(port, new Reservation(portFile, channel, lock));
+      return true;
+    } catch (IOException | OverlappingFileLockException e) {
+      return false;
+    }
+  }
+
+  /**
+   * @param port the port to release
+   */
+  public void release(int port) {
+    Reservation r = mReserved.remove(port);
+    if (r != null) {
+      r.mFile.delete();
```

Comment:
this doesn't throw anything? And, we can delete a file, and then close its channel?

Reply:
no, on failure it just returns `false`. In that case, we should still release the lock, and the file doesn't do any harm. Added a comment

Reply:
I've also added a test that we can call `delete` successfully while holding the lock